xfn-gotemplate/main.go

```diff
@@ -0,0 +1,235 @@
+		return
+	}
+
+	// Render the templates.
+	buf := &bytes.Buffer{}
+	if err := tpl.Execute(buf, vals); err != nil {
+		failFatal(out, errors.Wrap(err, "cannot execute template"))
+		return
+	}
+
+	// Parse the rendered manifests.
+	var objs []*unstructured.Unstructured
+	decoder := kyaml.NewYAMLOrJSONDecoder(bytes.NewBufferString(buf.String()), 1024)
+	for {
+		u := &unstructured.Unstructured{}
+		if err := decoder.Decode(&u); err != nil {
+			if err == io.EOF {
+				break
+			}
+			failFatal(out, errors.Wrap(err, "cannot decode manifest"))
+			return
+		}
+		if u != nil {
+			objs = append(objs, u)
+		}
+	}
+
+	// Convert the rendered manifests to a list of desired resources.
+	desiredResources := make(map[string]fnv1alpha1.DesiredResource, len(objs))
+	for _, obj := range objs {
+		res := desiredResources[obj.GetName()]
+		paved := fieldpath.Pave(obj.Object)
+		// We have two meta types to make connection details and readiness checks configurable in the templates.
+		if obj.GetAPIVersion() == "meta.gotemplate.xfn.upbound.io/v1alpha1" {
+			switch obj.GetKind() {
+			case "ConnectionDetails":
+				if err = paved.GetValueInto("connectionDetails", &res.ConnectionDetails); err != nil {
+					failFatal(out, errors.Wrap(err, "cannot get connection details"))
+					return
+				}
+			case "ReadinessChecks":
+				if err = paved.GetValueInto("readinessChecks", &res.ReadinessChecks); err != nil {
+					failFatal(out, errors.Wrap(err, "cannot get readiness checks"))
+					return
+				}
+			default:
+				failFatal(out, fmt.Errorf("unknown kind %s", obj.GetKind()))
+				return
+			}
+			desiredResources[obj.GetName()] = res
+			continue
+		}
+		b, err = obj.MarshalJSON()
+		if err != nil {
+			failFatal(out, errors.Wrap(err, "cannot marshal manifest"))
+			return
+		}
+		res.Resource = runtime.RawExtension{
+			Raw:    b,
+			Object: obj,
+		}
+		desiredResources[obj.GetAnnotations()["gotemplate.xfn.upbound.io/resource-name"]] = res
+	}
+
+	// Merge the desired resources with the existing ones.
+	desired := in.Desired
+	for name, res := range desiredResources {
+		found := false
+		for i, r := range desired.Resources {
+			if r.Name == name {
+				desired.Resources[i].Resource = res.Resource
+				desired.Resources[i].ConnectionDetails = res.ConnectionDetails
+				desired.Resources[i].ReadinessChecks = res.ReadinessChecks
+
+				// This is an already existing resource, we just update it without appending again.
+				found = true
+				break
+			}
+		}
+		if found {
+			continue
+		}
+		desired.Resources = append(desired.Resources, fnv1alpha1.DesiredResource{
+			Name:              name,
+			Resource:          res.Resource,
+			ConnectionDetails: res.ConnectionDetails,
+			ReadinessChecks:   res.ReadinessChecks,
+		})
+	}
+
+	// Marshal and write the output.
+	out.Desired = desired
+	b, err = yaml.Marshal(out)
+	if err != nil {
+		failFatal(out, errors.Wrap(err, "cannot marshal output"))
```

Comment:
```suggestion
		failFatal(fnv1alpha1.FunctionIO{}, errors.Wrap(err, "cannot marshal output"))
```

If the marshalling fails, I am pretty sure the `failFatal` will also fail - since it marshalls the input function. 😉